### PR TITLE
Be smarter about when we try to run scripting hooks

### PR DESCRIPTION
**Pull request by PhantomHoover**
Don't even bother trying if no scripts are even registered to run on them.

### CODE REVIEW

**Comment by wookieejedi** [comment]
Works as expected in tests with script-heavy mods such as FotG. Also, seems like it's 5-8 FPS faster in FotG's internal heavy stress test missions.

**Comment by PhantomHoover** [comment]
"But Hoover", I hear you say, "how can you be sure these checks aren't going to affect existing behaviour!" Good question; I believe these two checks:
https://github.com/scp-fs2open/fs2open.github.com/blob/master/code/scripting/scripting.cpp#L607
https://github.com/scp-fs2open/fs2open.github.com/blob/master/code/scripting/scripting.cpp#L622

mean that `ConditionedHook::IsOverride` and `ConditionedHook::Run` will return before doing anything important if the `action` passed to them doesn't match at least one `ConditionedHook`'s `action_type`. If this is the case the corresponding entry in `ActiveActions` should be set to true and so everything will be executed as before, and otherwise no scripts or other observable behaviour could've run anyway.

**Comment by JohnAFernandez** [comment]
Looking at the changes, it's mostly simple changes, even though it's in a lot of files.

**Comment by PhantomHoover** [comment]
Oh, also, for posterity, here's the averaged framerates of 10 runs each of Icarus with and without this change, on my system:
plot

**Review by z64555 — CHANGES_REQUESTED** [review]
LGTM, could maybe double check the bracket/bracing style if you want to be thorough.

I do have an objection to seeding the RNG with a static number in freespace.cpp:1600, however.  I do agree that there should be an option to use a static or user defined seed for benchmarking, but I don't quite agree to having it as a default option, and this tidbit is logically different from the rest of the PR.

File: code/scripting/scripting.h [hunk 1]
```diff
@@ -152,10 +152,9 @@ struct script_action
 
 class ConditionedHook
 {
-private:
+public:
```
Comment (asarium) [inline code/scripting/scripting.h:155]: Why is this necessary to be public now? I don't see a reason in the code to change that.

File: code/scripting/scripting.h [hunk 2]
```diff
@@ -243,6 +242,10 @@ class script_state
 	void ParseGlobalChunk(ConditionalActions hookType, const char* debug_str=nullptr);
 	bool ParseCondition(const char *filename="<Unknown>");
 	void AddConditionedHook(ConditionedHook hook);
+	// ActiveActions lets code that might run scripting hooks know whether any scripts are even registered for it.
+	// AssayActions is responsible for keeping it up to date.
+	bool ActiveActions[ConditionalActions::CHA_LAST+1];
```
Comment (asarium) [inline code/scripting/scripting.h:247]: The check if there is an active action should be put into a function which can then use this data structure internally. That will make it easier to refactor the internals of the scripting system later if that is necessary.

File: freespace2/freespace.cpp [hunk 3]
```diff
@@ -1596,8 +1596,8 @@ void game_init()
 
 	Game_current_mission_filename[0] = 0;
 
-	// seed the random number generator
-	Random::seed(static_cast<unsigned int>(time(nullptr)));
+	// seed the random number generator, using a static seed if benchmarking to encourage consistency
+	Random::seed(Cmdline_benchmark_mode ? 2335 : static_cast<unsigned int>(time(nullptr)));
```
Comment (asarium) [inline freespace2/freespace.cpp:1600]: I agree with @z64555, this should be in its own PR since it has nothing to do with the scripting changes.

**Review by asarium — CHANGES_REQUESTED** [review]
Seems to be a sensible change, just some style things I noticed.

File: code/scripting/scripting.h (hunk 1 above)
Comment (PhantomHoover) [reply code/scripting/scripting.h:155]: Because AssayActions needs to be able to iterate over the Actions vector, but it isn't a ConditionedHook member function. Conditions doesn't strictly need to be private but I felt it was cleaner just to take the honest route and make the whole thing public.

File: freespace2/freespace.cpp (hunk 3 above)
Comment (PhantomHoover) [reply freespace2/freespace.cpp]: I've removed this. In all honesty it wasn't even producing more consistent benchmarks.

File: code/scripting/scripting.h (hunk 2 above)
Comment (PhantomHoover) [reply code/scripting/scripting.h]: I've refactored everything to do it this way.

**Review by z64555 — APPROVED** [review]

File: code/scripting/scripting.h (hunk 1 above)
Comment (asarium) [reply code/scripting/scripting.h:155]: Ok, that is a valid point. In that case, please just make it a `struct` and remove the `public` altogether.

Comment (MageKing17) [reply code/scripting/scripting.h:155]: The general convention throughout the codebase is that, if it has methods, it's a `class`.

File: code/scripting/scripting.cpp [hunk 4]
```diff
@@ -1137,10 +1141,34 @@ bool script_state::ParseCondition(const char *filename)
...
+	}
+
+	for (const auto &hook : ConditionalHooks) {
+		for (const auto &action : hook.Actions) {
+			if (action.action_type >= -1 && action.action_type <= CHA_LAST) {
+				ActiveActions[action.action_type] = true;
+			}
+		}
+	}
+}
```
Comment (Goober5000) [inline code/scripting/scripting.cpp:1166]: Why does this function clear the entire array and then re-populate it whenever it's called?  Why not initialize the array when the script system is initialized, and then just toggle an action to true when a hook is added?

Comment (EatThePath) [reply code/scripting/scripting.cpp:1166]: seems like this would be fine if this only gets called once, after all the script tables have been loaded? I'm failing right now to imagine  a case where this needs to be done more than once.

Comment (PhantomHoover) [reply code/scripting/scripting.cpp:1166]: It also had to be run in AddConditionedHook:

https://github.com/scp-fs2open/fs2open.github.com/pull/3538/files#diff-8763caafc296d3dd73ccf6d72b40529289c3908b57fa9f06da1fc26ac840f547R1146

I also added it to the Clear method on script_system, but that was probably overkill as I think that's one of those dumb cleanup functions we have that only actually gets called when the game is about to exit.

As for why I did it this way, I just found it easier to write a function that would guarantee the active action lookup was accurate and then drop it in between any code that might add new conditional hooks and its return to the scripting system, versus fiddling around with various individual sites that might add hooks. It may look a little inefficient but consider that script_state::RunCondition iterates over every single conditional hook and does even more stuff than this hundreds or thousands of times per frame. AssayActions is probably run a few dozen times on engine startup at most.

File: code/scripting/scripting.h (hunk 1 above)
Comment (asarium) [reply code/scripting/scripting.h:155]: Alright I guess.

**Review by asarium — APPROVED** [review]
